hw/drivers/crypto/crypto_da1469x/src/crypto_da1469x.c

```diff
@@ -0,0 +1,232 @@
+#include <os/mynewt.h>
+#include <mcu/da1469x_clock.h>
+#include <crypto/crypto.h>
+#include <crypto_da1469x/crypto_da1469x.h>
+
+static struct os_mutex gmtx;
+
+#define VALID_AES_KEYLEN(x) (((x) == 128) || ((x) == 192) || ((x) == 256))
+
+static bool
+da1469x_has_support(struct crypto_dev *crypto, uint8_t op, uint16_t algo,
+                    uint16_t mode, uint16_t keylen)
+{
+    (void)op;
+
+    if (algo != CRYPTO_ALGO_AES || !VALID_AES_KEYLEN(keylen)) {
+        return false;
+    }
+
+    switch (mode) {
+    case CRYPTO_MODE_ECB:
+    case CRYPTO_MODE_CBC:
+    case CRYPTO_MODE_CTR:
+        return true;
+    }
+
+    return false;
+}
+
+static uint32_t
+da1469x_crypto_op(struct crypto_dev *crypto, uint8_t op, uint16_t algo,
+                  uint16_t mode, const uint8_t *key, uint16_t keylen,
+                  uint8_t *iv, const uint8_t *inbuf, uint8_t *outbuf,
+                  uint32_t len)
+{
+    uint8_t i;
+    uint8_t iv_save[AES_BLOCK_LEN];
+    unsigned int inc;
+    unsigned int carry;
+    unsigned int v;
+    uint32_t mask;
+
+    if (!da1469x_has_support(crypto, op, algo, mode, keylen)) {
+        return 0;
+    }
+
+    if (mode == CRYPTO_MODE_CBC && op == CRYPTO_OP_DECRYPT) {
+        memcpy(iv_save, &inbuf[len-AES_BLOCK_LEN], AES_BLOCK_LEN);
+    }
+
+    os_mutex_pend(&gmtx, OS_TIMEOUT_NEVER);
+    da1469x_clock_amba_enable(CRG_TOP_CLK_AMBA_REG_AES_CLK_ENABLE_Msk);
+
+    /*
+     * - enable AES / disable HASH
+     * - no more data in
+     * - disable IRQ
+     * - write out all data
+     */
+    AES_HASH->CRYPTO_CTRL_REG &=
+        ~(AES_HASH_CRYPTO_CTRL_REG_CRYPTO_HASH_SEL_Msk |
+          AES_HASH_CRYPTO_CTRL_REG_CRYPTO_MORE_IN_Msk |
+          AES_HASH_CRYPTO_CTRL_REG_CRYPTO_IRQ_EN_Msk |
+          AES_HASH_CRYPTO_CTRL_REG_CRYPTO_OUT_MD_Msk);
+
+    /* AES algorithm, there's no need to check algo here! */
+    AES_HASH->CRYPTO_CTRL_REG &= ~AES_HASH_CRYPTO_CTRL_REG_CRYPTO_ALG_MD_Msk;
+
+    /*
+     * Note: DS also mentions value 1 for ECB.
+     */
+
+    /* Select mode */
+    AES_HASH->CRYPTO_CTRL_REG &= ~AES_HASH_CRYPTO_CTRL_REG_CRYPTO_ALG_Msk;
+    switch (mode) {
+    case CRYPTO_MODE_ECB:
+        AES_HASH->CRYPTO_CTRL_REG |= 0 << AES_HASH_CRYPTO_CTRL_REG_CRYPTO_ALG_Pos;
+        break;
+    case CRYPTO_MODE_CBC:
+        AES_HASH->CRYPTO_CTRL_REG |= 2 << AES_HASH_CRYPTO_CTRL_REG_CRYPTO_ALG_Pos;
+        break;
+    case CRYPTO_MODE_CTR:
+        AES_HASH->CRYPTO_CTRL_REG |= 3 << AES_HASH_CRYPTO_CTRL_REG_CRYPTO_ALG_Pos;
+        break;
+    }
+
+    /*
+     * Note: iv writing to MREGx might require os_bswap_32!
+     */
+
+    switch (mode) {
+    case CRYPTO_MODE_CBC:
+    case CRYPTO_MODE_CTR:
+        AES_HASH->CRYPTO_MREG0_REG = ((uint32_t *)iv)[0];
+        AES_HASH->CRYPTO_MREG1_REG = ((uint32_t *)iv)[1];
+        AES_HASH->CRYPTO_MREG2_REG = ((uint32_t *)iv)[2];
+        AES_HASH->CRYPTO_MREG3_REG = ((uint32_t *)iv)[3];
+    }
+
+    /*
+     * Note: DS also mentions value 3 for 256-bit key.
```

Comment:
Based on my work during encrypted flash driver development, I made the following observations
        /*  
         * XXX: Datasheet for da1469x Rev 2.0 (CFR0011-120-00)
         * indicates that a value of 2 and 3 correspond to
         * 256 bit AES Key, However, only 2 works.
         */